### ISSUE TITLE
mysql: upgrade to v8 for nextcloud v21

### ISSUE THREAD

**Issue by kyrofa** [nextcloud/nextcloud-snap#1626]
A warning on 20.0.6 states:

> MySQL version "5.7.33" is used. Nextcloud 21 will no longer support this version and requires MySQL 8 or higher.

We'll need to do that. This will be done in a manner that is completely transparent to users, so if you're coming across the issue as a user, don't worry-- you can safely ignore that message.

**Comment by drasbeck** [comment]
I get the same message with 20.0.7snap1

**Comment by kyrofa** [comment]
Yeah don't worry, it's not an error, it's a heads up. I promise we won't update you to v21 without it :smiley: .

**Comment by GambaJo** [comment]
> 
> 
> Yeah don't worry, it's not an error, it's a heads up. I promise we won't update you to v21 without it 😃 .

Great. I already startet to investigate how to do it. Now I will wait and drink tea ;)

**Comment by faarruuk** [comment]
hi, just a short question.. is there a way to remove this warning? because it makes my customer feeling insecure ;) and I don't have actually the time to run a DB upgrade..

**Comment by kyrofa** [comment]
Not that I know of, this is embedded in upstream Nextcloud. They've pulled stuff like this before where they sent a notification saying "there's a security vuln in nginx, please upgrade immediately!" and of course... the snap doesn't even use nginx.

**Comment by szaimen** [comment]
@faarruuk you might be able to hide those warnings with the custom_css app if it is really important to you.

**Comment by edvaldsula** [comment]
Hi,

Basically, we do not have to do anything from our side? The upgrade of MySQL will be done automatically from the Nextcloud when it will be updated to Nextcloud 21. So, when it is planned to be done?

**Comment by kyrofa** [comment]
> Basically, we do not have to do anything from our side?

That's correct.

> So, when it is planned to be done?

When I'm able. Sooner if someone else wants to take a crack at it.

**Comment by selfroasted** [comment]
I couldn't have thought of a better result in my DDG search for this issue. After getting this same error, I landed here, read that it was going to be taken care of without the need for me to really do anything except continue my backups, and am going to continue my evening. However, I am lifting my drink to cheers you fine folks behind this work. Thank you.

**Comment by caseydiers** [comment]
Observed here as well.  Look forward to an update to resolve this.  Wish I was qualified to help take care of it!  Thanks @kyrofa and whomever else is involved.  We appreciate you!

**Comment by Lotze83** [comment]
Thanks for the snap! i wait next update to up nextcloud to 21 :)

**Comment by vawaver** [comment]
Guys, have some rumours about the release date of SNAP 21?

**Comment by ftpmorph** [comment]
Hey, first off great news this will be sorted "automagically" as it were.

I updated MySQL on my VPS as soon as I saw the error just because it's a new install I figure it's best to have everything up to date now, rather than later when I have to worry about losing data.

The problem I'm having is I upgraded to MySQL 8.0.23 successfully in Ubuntu:

> apt policy mysql-server
> mysql-server:
>   Installed: 8.0.23-1ubuntu20.04

...and yet this has no effect on Nextcloud, which still gives me that error.

I take it this is because the actual version of MySQL being used is inside the snap, so the version I have installed on Ubuntu itself makes no difference here?

Does this mean I can uninstall MySQL from Ubuntu safely without breaking anything in Nextcloud? Because if it's not being used it's just wasting resources.

Cheers.

**Comment by kevinschachtschneider** [comment]
@ftpmorph 

> Does this mean I can uninstall MySQL from Ubuntu safely without breaking anything in Nextcloud

This is correct, if nothing else on your system uses the MySQL you installed via apt then it is safe to remove because, as you mentioned, the snap uses a bundled version of MySQL that only Nextcloud Snap uses.

**Comment by rorybellows** [comment]
Can I just verify then...if I installed Nextcloud via snap then it will automatically update to v21 and MySQL 8 right?

**Comment by kyrofa** [comment]
That's correct @rorybellows.

**Comment by tawissus** [comment]
Is there a schedule yet or could the upgrade still take some time? In the meantime there will surely be the next updates of version 20 or?

**Comment by SireAngelus** [comment]
Hi! do we have an ETA? it's almost April. Otherwise i might be willing to have a crack at it but i'm a total noob on how to create snap packages.

**Comment by JMoVS** [comment]
@kyrofa Really love and appreciate your work! So far it's been super smooth sailing with the snap (aside from some weird cashing things because of snap) - is this upgrade here proving to be more difficult? I'm eagerly awaiting Nextcloud 21 for its Whiteboard feature ;-)

**Comment by RayCulp** [comment]
Our hosting provider will not be providing support for MySQL 8 any time soon. Do I understand correctly that the best alternative is to switch to MariaDB or PostgreSQL?

**Comment by kyrofa** [comment]
Not if you're using the snap. It doesn't matter what your provider supports, at that point.

**Comment by kyrofa** [comment]
No ETA yet, by the way. I've started, but I need to finish repairing fence on the ranch so I can get the cows on my pasture, so we still have a little ways to go on this one.

**Comment by RayCulp** [comment]
> 
> 
> Not if you're using the snap. It doesn't matter what your provider supports, at that point.

I'm not currently using the snap and I'm not sure my hosting provider / solution supports snaps.

**Comment by kyrofa** [comment]
Ah, I see your edit. Yes, I agree with your assessment.

**Comment by RayCulp** [comment]
> 
> 
> Ah, I see your edit. Yes, I agree with your assessment.

Thanks!

**Comment by RayCulp** [comment]
I'm afraid my hosting provider got back to me with very bad news. They do not support MariaDB, PostgreSQL or Snaps, and they will not in the foreseeable future. The only alternative they can offer is to let me try a Linux VPS for free for a month to see if I can move my existing installation with all files and data that it already contains and try to get it up and running on that. 

Now, I have an MCSA certification but almost no experience with Linux servers. In my experience, you cannot simply set up a server and allow access to it from the Internet. You absolutely must harden it, or you are inviting disaster. Let's say I manage to move my existing Nextcloud to a VPS and get it running. I would not know where to start to harden Linux, and I feel extremely apprehensive about renting a VPS, installing Nextcloud and opening the doors wide for anyone on the Internet to try to hack it. 

What other options exist? Can anyone recommend hosting providers with inexpensive shared hosting solutions that fulfill the system requirements for NC 21?

**Comment by ph00lt0** [comment]
@RayCulp this is very off-topic. Please discuss this somewhere else. But one advice: do not run Nextcloud on shared hosting, this is a bad idea. Digital ocean has quite a few good guides on how to setup your VPS for nextcloud. But if this is too complicated for you, you might be better of with offerings of f.x.: https://www.hetzner.com/storage/storage-share or https://shop.hanssonit.se/product/nextcloud-vm-1-tb/

**Comment by JMoVS** [comment]
> No ETA yet, by the way. I've started, but I need to finish repairing fence on the ranch so I can get the cows on my pasture, so we still have a little ways to go on this one.

Sounds like a good work-life balance! ;-)

**Comment by RayCulp** [comment]
@ph00lt0 You're right, it's off-topic. I'll take the discussion someplace else. But before I go, two things:
1. I've been running Nextcloud on shared hosting for more than two years without any problems whatsoever. Why would anyone want to take on the added workload and complexity of administrating a VPS on top of administrating NC when shared hosting simply works. Because it doesn't scale up to 100,000 users? Pfff.
2. I just talked to Hetzner and Hansonit on the phone. Hetzner does not support MySQL 8. Hansonit recommends the VM.

**Comment by ivr588** [comment]
![image](https://user-images.githubusercontent.com/15389379/115964901-15e31800-a516-11eb-9efc-23b34e0c8284.png)

why has this happened. 
nextcloud instance works.

os is freebsd-12.2-p6. nextcloud is configured to use its own ecosystem, rather than freebsd ports.

**Comment by kyrofa** [comment]
Interesting, @ivr588. I don't know much about snaps on freebsd, but would you mind logging a new issue so we can make sure you're in a good place?

**Comment by ivr588** [comment]
> Interesting, @ivr588. I don't know much about snaps on freebsd, but would you mind logging a new issue so we can make sure you're in a good place? Please include the output of `sudo nextcloud.mysql-client --version`.

Hi, there's no exact snaps equivalent on freebsd. There are packages and ports. Packages are pre-compiled binaries like you'd get with apt on ubuntu. Ports are makefiles to build said packages. But all this is a red herring as I do not use freebsd's ports/packages ecosystem to manage my nextcloud instance as nextcloud has its own ecosystem. From the viewpoint of nextcloud it is a standalone instance which has apache24,php74 and mysql56-server in its environment.

My concern is that nextcloud detected the mysql environment as 56, and upgraded itself from nextcloud20 to nextcloud21 anyway. It's good that it's still working, but I'm worried right now that it's not working as it should and I'm just not noticing it.

I'll create a new issue and link it here.

**Comment by ivr588** [comment]
> Interesting, @ivr588. I don't know much about snaps on freebsd, but would you mind logging a new issue so we can make sure you're in a good place? Please include the output of `sudo nextcloud.mysql-client --version`.

https://github.com/nextcloud/server/issues/26742#

**Comment by Fuseteam** [comment]
> My concern is that nextcloud detected the mysql environment as 56, and upgraded itself from nextcloud20 to nextcloud21 anyway. It's good that it's still working, but I'm worried right now that it's not working as it should and I'm just not noticing it.

nextcloud 21 is out yet as a snap tho and it will not upgrade itself automatically if you are on the `20/stable channel`

**Comment by kyrofa** [comment]
> Hi, there's no exact snaps equivalent on freebsd

Ah, then yeah, you're in the wrong place, this repo is for the snap only.

**Comment by runil** [comment]
I stupidly upgraded to master build a while back and luckily the core features have been working for several months. 

I'm running into issues with some applications (eg. when trying to install the Mail-app, i get the error "There is no table with name 'nextcloud.oc_mail_messages' in the schema." ). I can see that even though I'm running Nextcloud 22, the server still reports that the MySQL-database is of version 5. 

Will the migration solution that's in the working allow users of the master-build to upgrade their MySQL-version?

**Comment by Fuseteam** [comment]
reminder that this repository is for the snap variant of nextcloud

**Comment by kyrofa** [comment]
Alright folks, we have #1710 up that could use some testing if anyone is up for it.

**Comment by kyrofa** [comment]
Last call for testing, I'm hoping to land this tomorrow.

**Comment by HvdK-zOS** [comment]
Hello Kyrofa,

on my Nextcloud server it worked.

:~$ mysql -V
mysql  Ver 8.0.23-0ubuntu0.20.10.1 for Linux on x86_64 ((Ubuntu))

Nextcloud is currently on 20.0.9 

And no warning any more ....

Thanks !

**Comment by kyrofa** [comment]
Okay landed, it will be contained in 20.0.9snap2. Keep an eye on #1721; once it lands folks will start automatically updating.

**Comment by kyrofa** [comment]
Sorry for the delay on this folks. I always test new releases on my own production instance before rolling it to you, and I thought this upgrade hung on me. Turns out I was wrong, it just took a really long time (like 5 minutes). I was used to the mysql upgrader between 5.7 releases, apparently upgrading  to v8 just takes a while on a big database. My instance is massive though, so that probably won't be the average case, but I wanted to give you a heads up. I'm rolling this out now. If you feel like it's taking a while, watch `/var/snap/nextcloud/current/logs/mysql_errors.log` and you'll see it happening:

```
2021-05-14T13:31:49.037582Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
2021-05-14T13:31:54.867120Z 1 [System] [MY-013577] [InnoDB] InnoDB initialization has ended.
2021-05-14T13:32:06.951561Z 0 [System] [MY-011323] [Server] X Plugin ready for connections. Socket: /tmp/mysqlx.sock
2021-05-14T13:33:17.206627Z 2 [System] [MY-011003] [Server] Finished populating Data Dictionary tables with data.
2021-05-14T13:33:47.130278Z 5 [System] [MY-013381] [Server] Server upgrade from '50700' to '80024' started.
2021-05-14T13:36:03.356780Z 5 [System] [MY-013381] [Server] Server upgrade from '50700' to '80024' completed.
```

**Comment by caseydiers** [comment]
Thanks Kyle. Refreshed the snap and everything worked perfectly. Start to finish was less than a minute on a relatively small deployment.

**Comment by rorybellows** [comment]
> Thanks Kyle. Refreshed the snap and everything worked perfectly. Start to finish was less than a minute on a relatively small deployment.

Can you explain how you refreshed the snap please?

**Comment by caseydiers** [comment]
sudo snap refresh

**Comment by kyrofa** [comment]
> Can you explain how you refreshed the snap please?

Note that you don't have to do anything unless you want to upgrade _right now._ It'll upgrade on its own within a few hours unless you've tweaked the update settings for snaps.

**Comment by tawissus** [comment]
Version: 7.4.18 **✓**
Version: 8.0.24 **✓**
**20.0.9snap2 ✓**

Thank you for all the work! So far everything is working fine.

**Comment by DeirdreYoung** [comment]
PHP : Version: 7.4.18
MySQL : Version: 8.0.24
Status.php says : version "20.0.9.1" - and I'm using the snap package.
Admin overview says : "All checks passed" - I'm very happy to see this.

WebDAV works fine with Windows Explorer, I'm able to see all my files (and rename, as a test)

Kyrofa, thank you for your hard work, and everyone else pitching in who I don't know.

**Comment by Fuseteam** [comment]
@DeirdreYoung how did you resolve the imagick "issue"?

**Comment by szaimen** [comment]
> how did you resolve the imagick "issue"?

Disable the theming app and the issue will be gone.

**Comment by DeirdreYoung** [comment]
I confirm that I do not have theming enabled and didn't hit the imagick issue.

**Comment by rorybellows** [comment]
My Nextcloud just upgraded to version 21.0.1 and then now I'm getting this message:

There are some warnings regarding your setup.

    Your installation has no default phone region set. This is required to validate phone numbers in the profile settings without a country code. To allow numbers without a country code, please add "default_phone_region" with the respective ISO 3166-1 code ↗ of the region to your config file.

What's going on?

**Comment by d4n13Lh** [comment]
Look up your [ISO 3166-1](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2#Officially_assigned_code_elements) country code and then use
`sudo nextcloud.occ config:system:set default_phone_region --value=CH`  (replace CH with yours)
to store it in the config.

**Comment by rorybellows** [comment]
> 
> 
> Look up your [ISO 3166-1](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2#Officially_assigned_code_elements) country code and then use
> `sudo nextcloud.occ config:system:set default_phone_region --value=CH` (replace CH with yours)
> to store it in the config.

Worked like a charm!  Thank you!!!

**Comment by kyrofa** [comment]
I sure wish the Nextcloud developers would stop throwing random suggestions into that section that look like warnings.

**Comment by Fuseteam** [comment]
yeah agreed, its not like we __all__ use phone numbers in the system. i don't even have my number saved into it and even if i did if i want international calls i'll just add my country code lol
